### PR TITLE
Security Level subsystem

### PR DESCRIPTION
**Pull request by Arkatos1**
## About The Pull Request

This PR introduces a new subsystem - Security Level and improves code surrounding it.

Special security level code interactions - like changing lights for firealarms accordingly to a newly set security level are now handled via signals instead of iterating every machine in existence each time security level changes. Escape pods will now properly re-lock if security level should drop below red.

This also opens way for more interesting security level interactions, like machines changing or restricting behavior accordingly to a security level, or high security level being time limited and requiring re-confirmation every set period of time.

## Why It's Good For The Game

Cleaner code.

## Changelog
:cl: Arkatos
refactor: Added security level subsystem.
fix: Escape pods will now re-lock again properly if security level drops below red.
/:cl:

### CODE REVIEW

File: code/game/machinery/doors/door.dm [hunk 1]
```diff
@@ -93,6 +94,23 @@
...
+ * * new_level The new security level that is in effect
+ */
+/obj/machinery/door/proc/check_security_level(datum/source, new_level)
+	SIGNAL_HANDLER
+
+	if(new_level <= SEC_LEVEL_BLUE)
+		return
+	if(!red_alert_access)
+		return
+	visible_message("<span class='notice'>[src] whirrs as it automatically lifts access requirements!</span>")
```
Comment (Mothblocks) [inline code/game/machinery/doors/door.dm:111]:
```suggestion
	visible_message("<span class='notice'>[src] whirrs as [p_they()] automatically lifts access requirements!</span>")
```

File: code/modules/shuttle/emergency.dm [hunk 2]
```diff
@@ -603,6 +607,23 @@
...
+ */
+/obj/machinery/computer/shuttle/pod/proc/check_lock(datum/source, new_level)
+	SIGNAL_HANDLER
+
+	if(obj_flags & EMAGGED)
+		return
+	if(new_level >= SEC_LEVEL_RED)
+		locked = FALSE
+	else
+		locked = TRUE
```
Comment (Mothblocks) [inline code/modules/shuttle/emergency.dm:625]:
```suggestion
	locked = new_level < SEC_LEVEL_RED
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Two nit-picks.

**Review by Mothblocks — APPROVED** [review]

File: code/__DEFINES/subsystems.dm [hunk 3]
```diff
@@ -113,6 +113,7 @@
 #define INIT_ORDER_VIS 80
 #define INIT_ORDER_ACHIEVEMENTS 77
 #define INIT_ORDER_RESEARCH 75
+#define INIT_ORDER_SECURITY_LEVEL 75
```
Comment (coiax) [inline code/__DEFINES/subsystems.dm:116]: Does the security level subsystem depend on being before or after another subsystem? At a glance, I don't see any dependencies on any other subsystems.

If there are no dependencies, then this should just use the default place in the init order.

File: code/game/machinery/doors/door.dm (hunk 1 above)
Comment (coiax) [reply code/game/machinery/doors/door.dm]: Think you'll need to add an `[p_s()]` to the s on lifts too?

Comment (maxymax13) [reply code/game/machinery/doors/door.dm]: i know this is for double doors but i can't help but imagine "bob the Man Door, half man half door"

File: code/__DEFINES/subsystems.dm (hunk 3 above)
Comment (Arkatos1) [reply code/__DEFINES/subsystems.dm]: Changed to default place.

File: code/game/machinery/doors/door.dm (hunk 1 above)
Comment (Arkatos1) [reply code/game/machinery/doors/door.dm]: Added `[p_s()]` to both.

Comment (Mothblocks) [reply code/game/machinery/doors/door.dm]: I had no idea `p_s` was a thing. 😛

**Pull request merged**